Checkpoint/connectors/checkpoint_harmony.py

```diff
@@ -0,0 +1,153 @@
+"""Contains CheckpointHarmony connector."""
+import asyncio
+import time
+from datetime import datetime, timedelta, timezone
+from typing import Any, Optional
+
+import orjson
+from dateutil.parser import isoparse
+from loguru import logger
+from sekoia_automation.aio.connector import AsyncConnector
+from sekoia_automation.connector import DefaultConnectorConfiguration
+from sekoia_automation.storage import PersistentJSON
+
+from client.http_client import CheckpointHttpClient
+from connectors import CheckpointModule
+
+from .metrics import EVENTS_LAG, FORWARD_EVENTS_DURATION, OUTCOMING_EVENTS
+
+
+class CheckpointHarmonyConfiguration(DefaultConnectorConfiguration):
+    """The CheckpointHarmony configuration."""
+
+    ratelimit_per_minute: int = 60
+    chunk_size: int = 1000
+
+
+class CheckpointHarmonyConnector(AsyncConnector):
+    """The CheckpointHarmony connector."""
+
+    name = "CheckpointHarmonyConnector"
+    module: CheckpointModule
+    configuration: CheckpointHarmonyConfiguration
+
+    _checkpoint_client: CheckpointHttpClient | None = None
+
+    def __init__(self, *args: Any, **kwargs: Optional[Any]) -> None:
+        """Init CheckpointHarmony."""
+
+        super().__init__(*args, **kwargs)
+        self.context = PersistentJSON("context.json", self._data_path)
+
+    @property
+    def last_event_date(self) -> datetime:
+        """
+        Get last event date.
+
+        Returns:
+            datetime:
+        """
+        now = datetime.now(timezone.utc)
+        one_hour_ago = (now - timedelta(hours=1)).replace(microsecond=0)
+
+        with self.context as cache:
+            last_event_date_str = cache.get("last_event_date")
+
+        # If undefined, retrieve events from the last 1 hour
+        if last_event_date_str is None:
+            return one_hour_ago
+
+        # Parse the most recent date seen
+        last_event_date = isoparse(last_event_date_str).replace(microsecond=0)
+
+        # We don't retrieve messages older than 1 hour
+        if last_event_date < one_hour_ago:
+            return one_hour_ago
+
+        return last_event_date
+
+    def get_checkpoint_client(self) -> CheckpointHttpClient:
+        """Get CheckpointHttpClient."""
+        if self._checkpoint_client is None:
+            self._checkpoint_client = CheckpointHttpClient(
+                client_id=self.module.configuration.client_id,
+                secret_key=self.module.configuration.secret_key,
+                auth_url=self.module.configuration.authentication_url,
+                base_url=self.module.configuration.base_url,
+            )
+
+        return self._checkpoint_client
+
+    async def get_checkpoint_harmony_events(self) -> list[str]:
```

Comment:
How does this method handle pages of events when interacting with the API? I can't find how the client iterates over the pages in this method nor in the `get_harmony_mobile_alerts` method.

Reply:
Post an update in DM